File: problems/longest-substring-without-repeating-characters/solution3.py
def length_of_longest_substring(s: str) -> int:
    if not s:
        return 0
    # unique characters
    chars = set()
    start = end = 0
    max_length = float('-inf')
    while end < len(s):
        if s[end] not in chars:
            chars.add(s[end])
            end += 1
            max_length = max(max_length, len(chars))
        else:
            chars.remove(s[start])
            start += 1
    return max_length

File: problems/longest-substring-without-repeating-characters/Solution3.java
import java.util.HashSet;

public class Solution3 {
  public static int lengthOfLongestSubstring(String s) {
    if (s.isEmpty()) {
      return 0;
    }
    // unique characters
    final var characters = new HashSet<Character>();

    var start = 0;
    var end = 0;
    var maxLength = Integer.MIN_VALUE;
    while (end < s.length()) {
      if (characters.add(s.charAt(end))) {
        maxLength = Math.max(maxLength, characters.size());
        end++;
      } else {
        characters.remove(s.charAt(start));
        start++;
      }
    }
    return maxLength;
  }
}
